File: new/739_daily_temperatures_stack.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        // idea: stack
        // have a stack to track pair of {temp, index}. 
        // this stack tracks temp that are lesser than current index. 
        // if the top of the stack is less than current index, get its
        // index and the difference is number of days until higher temp. 
        // while loop the stack with current temp and current index until 
        // we go through all the lesser temperatures than current temperatures, 
        // and update our until array which is our resulting arrays of number
        // of days until better temperature. 
        int n = temperatures.size();
        vector<int> until(n);
        stack<pair<int, int>> lower_temp;
        for (int i = 0; i < temperatures.size(); i++) {
            int current_temp = temperatures[i];
            while (!lower_temp.empty() && lower_temp.top().first < current_temp) {
                auto pair = lower_temp.top();
                lower_temp.pop();
                until[pair.second] = i - pair.second;
            }
            lower_temp.push({current_temp, i});
        }

        return until;
    }
    // time: O(n)
    // space: O(n)
};

File: new/python/739_daily_temperatures_stack.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # idea: stack
        # use stack to track value and index of temperatures less than current
        # temperature, so we can find the difference between the current temperature
        # and lesser temperatures. We pop until we clear all the lesser temperatures
        # in the stack and update the list of days until higher temperature.
        # If current temperature is greater than the top
        # temperature in stack we do the above.
        # if not we add the current temperature to the stack and wait till greater temp
        # arrives.
        until = [0] * len(temperatures)
        stack = []
        for curr_index, curr_temp in enumerate(temperatures):
            while stack and curr_temp > stack[-1][0]:
                lesser_temp, lesser_index = stack.pop()
                until[lesser_index] = curr_index - lesser_index
            stack.append([curr_temp, curr_index])
        return until
        # time: O(n)
        # space: O(n)
